Allow BUNDLE_GEMFILE to not be an absolute path

gxespino opened the pull request:
Hello - I was stalking the open issues and had time to implement this quick PR for your review. This is a direct fix to https://github.com/bundler/bundler/issues/5712.

I implemented the fix that @segiddins suggested in the comments, just added a test scenario. Please let me know if anything else is needed!

---

### What was the end-user problem that led to this PR?

The problem was that starting with Bundler 1.15.0, `Bundler.setup` fails with an `ArgumentError` raised from `Pathname#relative_path_from` when the following conditions are met:

- Bundler is in deployment mode; and
- A Gemfile is explicitly specified via the `BUNDLE_GEMFILE` environment variable; and
- That Gemfile is not an absolute path (e.g. `BUNDLE_GEMFILE=Gemfile`)

### Was was your diagnosis of the problem?

My diagnosis was that in `Bundler::SharedHelpers#default_gemfile`, the `Pathname` object being instantiated was not being expanded. So, in the case that `ENV["BUNDLE_GEMFILE"]` was not an absolute path, `Bundle.setup` would error out `Bundler.default_lockfile.relative_path_from(SharedHelpers.pwd)` since `Bundler.default_lockfile` derived itself from `Bundler.default_gemfile`.

### What is your fix for the problem, implemented in this PR?

My fix was to add `expand_path` to the `Pathname` object being returned in `Bundler::SharedHelpers#default_gemfile`

### Why did you choose this fix out of the possible options?

I chose this fix because Bundler was already deriving `Bundler::Sharedhelpers#root` with an `#expand_path`, so I wanted to follow that pattern.

segiddins commented on spec/bundler/shared_helpers_spec.rb:
```diff
@@ -121,6 +121,14 @@
+      before { ENV["BUNDLE_GEMFILE"] = "Gemfile" }
+
+      it "returns path of the bundle Gemfile" do
+        expect(subject.in_bundle?).to eq("#{bundled_app}/Gemfile")
```
please test that `subject.find_gemfile` returns this, as only the truthiness of `in_bundle?` is guaranteed.

gxespino commented:
Hello. I updated the PR with a higher level unit test as well as an integration spec. I also changed the solution to move `expand_path` further up the chain. 

Let me know if this is not desired and/or if any more changes are needed!

segiddins commented:
:100: this is perfect, thank you so much!

@bundlerbot r+

mattbrictson commented:
I noticed this didn't make it into 1.15.2. Will it be ported to 1.15.x at some point?

segiddins commented:
It'll definitely go into 1.16
